examples/nfc-st25tb-info.1
.Dd June 9, 2021
.Dt NFC-ST25TB-INFO 1 URM
.Sh NAME
.Nm nfc-st25tb-info
.Nd Read ISO-14443-B ST25TB* and legacy SR* cards
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
.Nm 
is a demonstration tool that try to dump the content of ISO-14443-B ST25TB* and legacy SR* cards.
.Pp
It includes, in theory, cards from STMicroelectronics, like ST25TB512-AC, ST25TB512-AT, ST25TB02K and ST25TB04K (at this time, not supported by Proxmark3). Legacy STMicroelectronics cards like SRT512, SRI512, or others SR* can also be dumped (supported by Proxmark3).
.Sh IMPORTANT
This example has been developed using PN533 USB hardware as initiator.
.Pp
It was tested with these cards:
.Pp
- ST25TB512-AC - (BE/Brussels/STIB ; AliExpress ones)
.br
- ST25TB512-AT - (FR/Lille/Ilevia ; FR/Reims/Citura)
.br
- SRT512 - legacy - (FR/Bordeaux/TBM)
.br
- SRI512 - legacy - (anonymous vending machine)
.Sh TODO
At this time, neither reading to file, neither loading from file is supported. Dealing with counters logic is dangerous and cannot be undo.
.Sh BUGS
Please report any bugs on the
.Em libnfc
issue tracker at:
.Em https://github.com/nfc-tools/libnfc/issues
.Sh LICENCE
-
.Em libnfc
is licensed under the GNU Lesser General Public License (LGPL), version 3.
.br
-
.Em libnfc-utils
and 
.Em libnfc-examples
are covered by the BSD 2-Clause license.
.br
-
.Em nfc-st25tb-info
is covered by Creative Commons Attribution 4.0 International (CC BY 4.0).
.Sh AUTHOR
.An Benjamin Delpy Aq benjamin@gentilkiwi.net
.Pp
This manual page was written by Benjamin Delpy.
It is licensed under the terms of the GNU GPL (version 2 or later).

NEWS
New in 1.7.1:

API Changes:

  * nfc_initiator_select_passive_target() provides defaults if pbtInitData=NULL
  * nfc_initiator_target_is_present() allow NULL pointer to tag

New in 1.7.0:

Drivers:

  * New PN532 over I2C driver, see contrib/libnfc/pn532_i2c_on_rpi.conf.sample

API Changes:

  * New function iso14443b_crc_append()

New in 1.7.0-rc7:

Drivers:

  * New PN532 over SPI driver, see contrib/libnfc/pn532_spi_on_rpi.conf.sample

API Changes:

  * Functions
    - nfc_initiator_target_is_present() & str_nfc_target():
      now take a pointer to nfc_target as argument
    - nfc_init(): upon malloc error, doesn't force exit() anymore
      so now you should test if context != NULL after nfc_init() call

New in 1.7.0-rc5:

API Changes:

  * Functions
    - New nfc_register_driver() function allowing to hook custom drivers.

New in 1.7.0-rc3:

API Changes:

  * Functions
    - Add timeout param to nfc_emulate_target()

New in 1.7.0-rc2:

Configuration:
  libnfc can now use a configuration file for special setups, or features
  activation. This file (/etc/nfc/libnfc.conf under GNU/Linux systems)
  supports already some keywords:
    - "allow_autoscan" to enable/disable device auto-detection feature;
    - "allow_intrusive_scan" to enable/disable intrusive auto-detection
      (ie. serial port probing);
    - "log_level" to select library verbosity;
    - "device.name" and "device.connstring" to define a user device,
      this is the recommended method if user has a not easily detectable
      device (ie. a serial one).
  It is also possible to define devices using dedicated configuration files and
  put them into device search directory (/etc/nfc/devices.d under GNU/Linux).
  Example for the OpenPCD2: create /etc/nfc/devices.d/openpcd2.conf with:
    name = "OpenPCD2"
    connstring = "pn532_uart:/dev/ttyACM0"
    optional = true
  The keyword "optional" does not mandate the device to be present always
  (it detects if the reader is indeed present before using it)

API Changes:

  * Types
    - New NFC_ESOFT error to handle software errors (allocations, pipe
      creation, etc.)

  * Functions
    - Remove nfc_get_default_device() function: the default device is now the
      first in nfc_list_devices() or could be open using NULL connstring with
      nfc_open() function.
    - New enum-to-string converter functions str_nfc_modulation_type() and
      str_nfc_baud_rate()
    - New str_nfc_target() to convert nfc_target struct into allocated string
    - New nfc_device_get_information_about() function to retreive some device's
      information
    - No more in/out function parameter: nfc_initiator_transceive_*() now
      take a constant size for Rx buffer
    - New nfc_initiator_target_is_present() to test is the previously selected
      target is available in the field
    - nfc_initiator_transceive_bytes() returns NFC_EMFCAUTHFAIL when AUTH
      command failed on a Mifare Classic
    - New nfc_initiator_init_secure_element() to initiate a connection with
      secure element (Only supported with a PN532 with SAM equipped)

New in 1.6.0-rc1:

API Changes:

  * Types
    - '_t' suffix removed from all types (e.g. nfc_device_t is now nfc_device)
    - All errors removed in flavour of NFC_EIO, NFC_EINVARG, NFC_EDEVNOTSUPP,
      NFC_ENOTSUCHDEV, NFC_EOVFLOW, NFC_ETIMEOUT, NFC_EOPABORTED, NFC_ENOTIMPL,
      NFC_ETGRELEASED, NFC_ERFTRANS, NFC_ECHIP and NFC_SUCCESS
    - nfc_device_desc_t replaced by nfc_connstring: libnfc now uses connection
      strings to describe a device
    - byte_t typedef removed, libnfc now uses uint8_t from C99
    - nfc_device is now an opaque type
    - nfc_properties replaces nfc_options

  * Functions
    - New nfc_get_default_device() function that allows to grab the connstring
      stored in LIBNFC_DEFAULT_DEVICE environnement variable or returns the
      first available device if not set
    - New nfc_device_get_connstring() accessor function to know the device
      connstring
    - New nfc_device_set_property_bool() function that replace nfc_configure()
    - New nfc_device_set_property_int() function to set integer property
    - nfc_device_name() renamed to nfc_device_get_name() for the sake of
      consistency
    - New nfc_device_get_last_error() function, an accessor to last error occured
    - Whole libnfc's functions now return 0 (NFC_SUCCESS) or positive value if
      appropriated on success and libnfc's error code on failure
    - nfc_connect(), nfc_disconnect() renamed to nfc_open(), nfc_close()
      respectively
    - Add 2 new functions: initialization and deinitialization functions:
      nfc_init() and nfc_exit()
    - New nfc_device_get_supported_modulation() and
      nfc_device_get_supported_baud_rate() functions

  * Dependencies
    - log4c is not anymore used for debugging facility. It was a bad choice,
      sorry for inconvenience.

New in 1.5.1:

API Changes

  * Types
    - Communication-level errors DEIO and DETIMEOUT are now know as ECOMIO,
      ECOMTIMEOUT respectively
    - Common device-level errors DEINVAL and DEABORT are now know as EINVALARG,
      EOPABORT respectively
    - New errors:  EFRAACKMISMATCH, EFRAISERRFRAME, EDEVNOTSUP and ENOTIMPL

  * Functions
    - nfc_abort_command() returns a boolean
    - timeout (struct timeval) pointer added to
      nfc_initiator_transceive_bytes(), nfc_target_send_bytes() and
      nfc_target_receive_bytes()
    - timed functions nfc_initiator_transceive_bytes_timed() and
      nfc_initiator_transceive_bits_timed() now takes uint32_t as cycles
      pointer
    - nfc_initiator_poll_targets() renamed to nfc_initiator_poll_target() and
      only return one target

New in 1.5.0:

Installed files
  - nfc-message.h have been removed, internal macros are not part of API.
  - New nfc-emulation.h file offers a middle level API to handle emulation (see
    nfc-emulate-forum-tag4 example)

API Changes

  * Types

examples/nfc-st25tb-info.c
/*-
 * Free/Libre Near Field Communication (NFC) library
 *
 * Libnfc historical contributors:
 * Copyright (C) 2009      Roel Verdult
 * Copyright (C) 2009-2013 Romuald Conty
 * Copyright (C) 2010-2012 Romain Tartière
 * Copyright (C) 2010-2013 Philippe Teuwen
 * Copyright (C) 2012-2013 Ludovic Rousseau
 * See AUTHORS file for a more comprehensive list of contributors.
 * Additional contributors of this file:
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *  1) Redistributions of source code must retain the above copyright notice,
 *  this list of conditions and the following disclaimer.
 *  2 )Redistributions in binary form must reproduce the above copyright
 *  notice, this list of conditions and the following disclaimer in the
 *  documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 * Note that this license only applies on the examples, NFC library itself is under LGPL
 *
 */

/**
 * @file nfc-st25tb-info.c
 * @brief Read ISO-14443-B ST25TB* and legacy SR* cards
 */

/*	Benjamin DELPY `gentilkiwi`
 *	https://blog.gentilkiwi.com
 *	benjamin@gentilkiwi.com
 *	Licence : https://creativecommons.org/licenses/by/4.0/
 *	Rely on : libnfc - https://github.com/nfc-tools/libnfc
 *	
 *	$ gcc -lnfc -o nfc-st25tb-info nfc-st25tb-info.c
 *	$ ./nfc-st25tb-info
*/
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <nfc/nfc.h>

bool get_info(const nfc_target *pnt, uint8_t *pnbBlock, uint8_t *pbnSystem);
bool get_block_at(nfc_device *pnd, uint8_t block, uint8_t value[4], bool bPrintIt);
void print_hex(const uint8_t *pbtData, const size_t szBytes);

int main(int argc, const char *argv[])
{
	nfc_context *context = NULL;
	nfc_device *pnd = NULL;
	nfc_target nt = {0};
	nfc_modulation nm = {NMT_ISO14443B, NBR_106};

	int res;
	uint8_t i, nbBlock, bnSystem;

	(void) argc;
	(void) argv;

	nfc_init(&context);
	if(context)
	{
		pnd = nfc_open(context, NULL);
		if(pnd)
		{
			res = nfc_initiator_init(pnd);
			if(res == NFC_SUCCESS)
			{
				printf("Reader  : %s - via %s\n  ...wait for card...\n", nfc_device_get_name(pnd), nfc_device_get_connstring(pnd));

				res = nfc_initiator_list_passive_targets(pnd, nm, &nt, 1);
				if(res == 0) // we don't really wanted a NMT_ISO14443B
				{
					nm.nmt = NMT_ISO14443B2SR; // we want a NMT_ISO14443B2SR, but needed to ask for NMT_ISO14443B before
					if (nfc_initiator_select_passive_target(pnd, nm, NULL, 0, &nt) > 0)
					{
						if(get_info(&nt, &nbBlock, &bnSystem))
						{
							printf("\nData    :\n");
							for(i = 0; i < nbBlock; i++)
							{
								get_block_at(pnd, i, NULL, 1);
							}
							get_block_at(pnd, bnSystem, NULL, 1);
						}
					}
				}
				else if(res > 0)
				{
					printf("ERROR - We got a NMT_ISO14443B ?\n");
				}
				else printf("ERROR - nfc_initiator_list_passive_targets: %i\n", res);
			}
			else printf("ERROR - nfc_initiator_init: %i\n", res);
			
			nfc_close(pnd);
		}
		else printf("ERROR - nfc_open\n");

		nfc_exit(context);
	}
	else printf("ERROR - nfc_init\n");

	return 0;
}

bool get_info(const nfc_target *pnt, uint8_t *pnbBlock, uint8_t *pbnSystem)
{
	bool bRet = false, bIsLegacy = false;
	const uint8_t *p;
	uint8_t chipId;

	*pnbBlock = 0x10;
	*pbnSystem = 0xff;
	
	if(pnt->nm.nmt == NMT_ISO14443B2SR)
	{
		printf("Target  : %s (%s)\nUID     : ", str_nfc_modulation_type(pnt->nm.nmt), str_nfc_baud_rate(pnt->nm.nbr));
		print_hex(pnt->nti.nsi.abtUID, sizeof(pnt->nti.nsi.abtUID));
		printf("\n");
		
		p = pnt->nti.nsi.abtUID;
		if(p[7] == 0xd0) // ST25TB* / SR*
		{
			bRet = true;
			chipId = p[5];
			printf("Manuf   : 0x%02x - %s\nChipId  : 0x%02x - ", p[6], (p[6] == 0x02) ? "STMicroelectronics" : "other", chipId);

			switch(chipId)
			{
				case 0x3f: // 00111111
					printf("ST25TB02K");
					*pnbBlock = 0x40;
					break;
				case 0x1f: // 00011111
					printf("ST25TB04K");
					*pnbBlock = 0x80;
					break;
				case 0x1b: // 00011011
					printf("ST25TB512-AC");
					break;
				case 0x33: // 00110011
					printf("ST25TB512-AT");
					break;
					
				default:
					chipId >>= 2;
					printf("legacy ? - 0x%02x - ", chipId);
					bIsLegacy = 1;
					switch(chipId)
					{
						case 0x02:
							printf("SR176");
							*pnbBlock = 0x0e;
							*pbnSystem = 0x0f;
							break;
						case 0x03:
							printf("SRIX4K");
							*pnbBlock = 0x80;
							break;
						case 0x04:
							printf("SRIX512");
							break;
						case 0x06:
							printf("SRI512");
							break;
						case 0x07:
							printf("SRI4K");
							*pnbBlock = 0x80;
							break;
						case 0x0c:
							printf("SRT512");
							break;
						default:
							bIsLegacy = 0;
							printf("unknown");
					}
			}
			
			printf("\nSerial  : 0x");
			if(bIsLegacy)
			{
				printf("%1x", p[5] & 0x03);
			}
			printf("%02x%02x%02x%02x%02x\n|usr blk: %hhu\n|sys blk: %hhu\n", p[4], p[3], p[2], p[1], p[0], *pnbBlock, *pbnSystem);
		}
		else printf("WARNI - Last byte of UID isn\'t 0xd0, but 0x%02x (not ST25TB / SR series?)\n", p[7]);
	}
	else printf("ERROR - not a NMT_ISO14443B2SR ?\n");
	
	return bRet;
}

bool get_block_at(nfc_device *pnd, uint8_t block, uint8_t value[4], bool bPrintIt)
{
	bool bRet = false;
	uint8_t tx[2] = {0x08, block}, rx[4];	
	int res;
	
	res = nfc_initiator_transceive_bytes(pnd, tx, sizeof(tx), rx, sizeof(rx), 0);
	if(res == 4)
	{
		bRet = true;

		if(value)
		{
			memcpy(value, rx, sizeof(rx));
		}

		if(bPrintIt)
		{
			printf("[%02x] ", block);
			print_hex(rx, sizeof(rx));
			printf("\n");
		}
	}
	else if(res > 0)
	{
		printf("ERROR - We got %i bytes?\n", res);
	}
	else printf("ERROR - nfc_initiator_transceive_bytes: %i\n", res);
	
	return bRet;
}

void print_hex(const uint8_t *pbtData, const size_t szBytes)
{
	size_t szPos;
	for (szPos = 0; szPos < szBytes; szPos++)
	{
		printf("%02x ", pbtData[szPos]);
	}
}